agent-bootstrap/src/main/java/io/opentelemetry/auto/bootstrap/instrumentation/decorator/HttpServerTracer.java

```diff
@@ -0,0 +1,283 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.opentelemetry.auto.bootstrap.instrumentation.decorator;
+
+import static io.opentelemetry.OpenTelemetry.getPropagators;
+import static io.opentelemetry.trace.Span.Kind.SERVER;
+import static io.opentelemetry.trace.TracingContextUtils.currentContextWith;
+import static io.opentelemetry.trace.TracingContextUtils.getSpan;
+
+import io.grpc.Context;
+import io.opentelemetry.OpenTelemetry;
+import io.opentelemetry.auto.config.Config;
+import io.opentelemetry.auto.instrumentation.api.MoreTags;
+import io.opentelemetry.auto.instrumentation.api.SpanWithScope;
+import io.opentelemetry.auto.instrumentation.api.Tags;
+import io.opentelemetry.context.propagation.HttpTextFormat;
+import io.opentelemetry.trace.Span;
+import io.opentelemetry.trace.SpanContext;
+import io.opentelemetry.trace.Status;
+import io.opentelemetry.trace.Tracer;
+import io.opentelemetry.trace.attributes.SemanticAttributes;
+import java.io.PrintWriter;
+import java.io.StringWriter;
+import java.lang.reflect.Method;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.concurrent.ExecutionException;
+import lombok.extern.slf4j.Slf4j;
+
+// TODO In search for a better home package
+@Slf4j
+public abstract class HttpServerTracer<REQUEST> {
+  public static final String SPAN_ATTRIBUTE = "io.opentelemetry.auto.span";
+
+  protected final Tracer tracer;
+
+  public HttpServerTracer() {
+    tracer = OpenTelemetry.getTracerProvider().get(getInstrumentationName(), getVersion());
+  }
+
+  public SpanWithScope startSpan(REQUEST request, Method origin, String originType) {
```

Comment:
I think I'd keep `SpanWithScope` out of this this new API, and have `HttpServerTracer` return and work only on spans.

`SpanWithScope` is primarily an auto-instrumentation detail to make it easier to pass them both between `@OnMethodEnter` and `@OnMethodExit` (in the future we could potentially get rid of it altogether using an extra `@Local` bytebuddy parameter, reducing memory allocations).

Reply:
I see your point, but don't know how do that. Take a look at line 73. This method currently both returns newly created span (or null) and new scope. And that scope incapsulates a span which may be different from the returned value. If we don't need this "rescoping" then things become easier.

Reply:
Maybe this rescoping should happen in the instrumentation that needs it? I don't think the rescoping is applicable to most of the http server instrumentation anyways?

Reply:
I have put this rescoping here because it looks like generic concern to me. I don't see how this can be instrumentation-specific.

Reply:
most of the http server instrumentations don't need this rescoping

Reply:
How do you decide that? I would think that all server instrumentations need that, why not?

Reply:
> E.g. in case of async servlet request processing we create span for incoming request in one thread, but actual request continues processing happens in another thread.
> Depending on servlet container implementation, this processing may again arrive into this method.
> E.g. Jetty handles async requests in a way that calls HttpServlet.service method twice.

Other http servers don't do this weird thing of calling the same entrypoint for a second time from another thread

Reply:
Current comment brings Jetty example only because I have personally seen this effect in Jetty. I have currently no idea if other containers can lead to the same effect. In `master` this rescoping logic lives in `Servlet3Advice`, which is servlet-container agnostic. I can move rescoping back there to preserve the current status quo.

Reply:
But coming back to you original request in this thread. Do you propose that tracers should only concern themselves with spans and not scopes? For me managing scope (aka "what is the current span") is a cross-cutting concern that we should, ideally, handle centrally. And not in every single instrumentation. If you don't want to do it here, then what common code will handle #465 ? I thought this tracer is a good place for that.

Reply:
My current thoughts:
* Have `HttpServerTracer.startSpan()` just create a new span (no conditional logic whether to create it or not, and no attaching to scope)
* Have a separate method that we call beforehand: `HttpServerTracer.shouldCreateSpan(...)` --> `boolean`, which can check the `SERVER` attribute in the current context
* Have another method that we call afterwards: `HttpServerTracer.withSpan(Span)` --> `Scope`, which can set the span in the `SERVER` attribute and attach the scope
* Ya, let's leave that rescoping logic only in `Servlet3Advice`, I don't think it's needed anywhere else

Some additional background:
* I don't particularly like `SpanWithScope`, it was basically just a workaround for passing both of those objects between `@OnMethodEnter` and `@OnMethodExit`, and at some point I'd like to get rid of `SpanWithScope`, instead adding a ByteBuddy `@Local` parameter to carry Span and Scope separately between `@OnMethodEnter` and `@OnMethodExit`
* I like being explicit about Scope/Context, and think it's ok to have a couple extra method calls to keep this more explicit

Reply:
Your suggestion looks good, I will try it. But it defeats one of my goals: to remove "constant call chains", where every instrumentation has to call exact same methods in exact same sequence.

But I will try first removing rescoping and will see how code will look like then.